File: polyglot/csharp/entropy_analyzer.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Text;
using System.Linq;
using System.Numerics;

namespace Polyglot.CSharp.EntropyAnalyzer
{
    /// <summary>
    /// Configuration for entropy analysis parameters.
    /// </summary>
    public sealed class EntropyConfig
    {
        public const int DefaultChunkSize = 65536; // 64KB chunks
        public const int MinFileSize = 1024;       // Minimum file size to analyze
        
        public int ChunkSize { get; set; } = DefaultChunkSize;
        public int WindowSize { get; set; } = 4096; // Sliding window for peak detection
        public double PeakThreshold { get; set; } = 7.5; // Bits per byte threshold
        
        public bool StreamMode { get; set; } = true;
    }

    /// <summary>
    /// Represents a single chunk's entropy analysis result.
    /// </summary>
    public sealed class ChunkResult
    {
        public long Offset { get; set; }
        public int Size { get; set; }
        public double Entropy { get; set; }
        public byte[] Data { get; set; }
        
        public bool IsHighEntropy => Entropy > 7.0;
    }

    /// <summary>
    /// Represents a peak entropy region detected in the file.
    /// </summary>
    public sealed class PeakRegion
    {
        public long StartOffset { get; set; }
        public int Length { get; set; }
        public double AverageEntropy { get; set; }
        public string LikelyCause { get; set; } // "Packer", "Encryption", etc.
        
        public static readonly PeakRegion Empty = new PeakRegion 
        { 
            StartOffset = -1, 
            Length = 0, 
            AverageEntropy = 0,
            LikelyCause = null 
        };

        public override string ToString() => 
            $"[{StartOffset:X8}] +{Length} @ {AverageEntropy:F2} bits/byte ({LikelyCause})";
    }

    /// <summary>
    /// Main entropy analyzer class.
    /// </summary>
    public sealed class EntropyAnalyzer : IDisposable
    {
        private readonly EntropyConfig _config;
        private readonly byte[] _buffer;
        private readonly List<ChunkResult> _chunks;
        
        public double TotalEntropy => CalculateTotalEntropy();

        public PeakRegion? HighEntropyPeak { get; private set; }

        public EntropyAnalyzer(EntropyConfig config = null)
        {
            _config = config ?? new EntropyConfig();
            _buffer = new byte[_config.ChunkSize];
            _chunks = new List<ChunkResult>();
        }

        /// <summary>
        /// Analyzes a file path and returns complete results.
        /// </summary>
        public AnalysisResult AnalyzeFile(string filePath)
        {
            if (string.IsNullOrEmpty(filePath))
                throw new ArgumentException("Path cannot be null or empty", nameof(filePath));

            var fileInfo = new FileInfo(filePath);
            
            if (!fileInfo.Exists || fileInfo.Length < _config.MinFileSize)
                return CreateEmptyResult(fileInfo.Name, 0, "File not found or too small");

            using (var stream = File.OpenRead(filePath))
            {
                return AnalyzeStream(stream, fileInfo.Name, fileInfo.Length);
            }
        }

        /// <summary>
        /// Analyzes an existing stream.
        /// </summary>
        public AnalysisResult AnalyzeStream(Stream stream, string name, long length)
        {
            if (stream == null || length <= 0)
                return CreateEmptyResult(name, 0, "Invalid stream");

            var result = new AnalysisResult
            {
                Name = name,
                Path = "", // Set by caller if needed
                TotalSize = length,
                OverallEntropy = 0.0,
                Chunks = new List<ChunkResult>(),
                Peaks = new List<PeakRegion>
            };

            long offset = 0;
            
            while (offset < length)
            {
                int toRead = _config.ChunkSize > 0 
                    ? Math.Min(_config.ChunkSize, (int)(length - offset))
                    : (int)Math.Min(65536, length - offset);

                if (toRead <= 0) break;

                stream.Position = offset;
                int bytesRead = stream.Read(_buffer, 0, toRead);

                if (bytesRead == 0) break;

                var chunkResult = CalculateChunkEntropy(offset, _buffer, bytesRead);
                result.Chunks.Add(chunkResult);

                // Track peaks
                if (chunkResult.Entropy > result.OverallEntropy && 
                   chunkResult.Entropy >= _config.PeakThreshold)
                {
                    result.OverallEntropy = chunkResult.Entropy;
                    
                    var peak = new PeakRegion
                    {
                        StartOffset = offset,
                        Length = bytesRead,
                        AverageEntropy = chunkResult.Entropy,
                        LikelyCause = DetectLikelyCause(chunkResult)
                    };

                    // Merge overlapping peaks
                    if (result.Peaks.Count > 0 && 
                       result.Peaks.Last().StartOffset + result.Peaks.Last().Length >= offset)
                    {
                        var lastPeak = result.Peaks[result.Peaks.Count - 1];
                        lastPeak.Length += bytesRead;
                        lastPeak.AverageEntropy = CalculateWeightedAverage(
                            lastPeak.AverageEntropy, bytesRead, 
                            chunkResult.Entropy);
                    }
                    else
                    {
                        result.Peaks.Add(peak);
                    }

                    HighEntropyPeak = peak;
                }

                offset += bytesRead;
            }

            // Calculate overall entropy from all chunks
            if (result.Chunks.Count > 0)
            {
                var totalBytes = result.Chunks.Sum(c => c.Size);
                double sumWeighted = result.Chunks.Sum(
                    c => c.Entropy * c.Size / totalBytes);
                
                result.OverallEntropy = sumWeighted;
            }

            return result;
        }

        /// <summary>
        /// Calculates entropy for a single chunk of data.
        /// </summary>
        private ChunkResult CalculateChunkEntropy(long offset, byte[] buffer, int size)
        {
            var counts = new byte[256];
            
            // Count byte frequencies
            foreach (var b in buffer.Take(size))
                counts[b]++;

            // Calculate Shannon entropy
            double entropy = 0.0;
            for (int i = 0; i < 256; i++)
            {
                if (counts[i] > 0)
                {
                    double p = counts[i] / size;
                    entropy -= p * Math.Log(p, 2);
                }
            }

            return new ChunkResult
            {
                Offset = offset,
                Size = size,
                Entropy = entropy,
                Data = buffer.Take(size).ToArray()
            };
        }

        /// <summary>
        /// Calculates total entropy across entire file.
        /// </summary>
        private double CalculateTotalEntropy()
        {
            if (_chunks.Count == 0) return 0;

            var totalBytes = _chunks.Sum(c => c.Size);
            double sumWeighted = _chunks.Sum(
                c => c.Entropy * c.Size / totalBytes);

            return sumWeighted;
        }

        /// <summary>
        /// Detects likely cause of high entropy region.
        /// </summary>
        private string DetectLikelyCause(ChunkResult chunk)
        {
            // Heuristics based on common packer characteristics
            if (chunk.Offset % 4096 == 0 && chunk.Entropy > 7.2)
                return "Possible UPX/Pack";

            if (chunk.Offset % 8192 == 0 && chunk.Entropy > 7.3)
                return "Possible Themida/VMProtect";

            // Check for common encryption patterns
            var byteCount = chunk.Data.Count(b => b < 32);
            double lowByteRatio = (double)byteCount / chunk.Size;

            if (lowByteRatio > 0.85)
                return "Possible Encryption/Compression";

            return "Unknown - High Entropy Region";
        }

        /// <summary>
        /// Calculates weighted average of two entropy values.
        /// </summary>
        private double CalculateWeightedAverage(double avg1, int size1, double avg2)
        {
            if (size1 == 0 || avg2 <= 0) return avg1;
            return (avg1 * size1 + avg2 * _config.WindowSize) / (size1 + _config.WindowSize);
        }

        /// <summary>
        /// Creates an empty result with default values.
        /// </summary>
        private AnalysisResult CreateEmptyResult(string name, long size, string reason = "")
        {
            return new AnalysisResult
            {
                Name = name,
                TotalSize = size,
                OverallEntropy = 0.0,
                Reason = reason
            };
        }

        /// <summary>
        /// Serializes results to JSON format compatible with packpeek.
        /// </summary>
        public string ToJson(AnalysisResult result)
        {
            var sb = new StringBuilder();
            
            sb.Append('{');
            sb.Append($"\"name\":\"{result.Name}\",");
            sb.Append($"\"size\":{result.TotalSize},");
            sb.Append($"\"overall_entropy\":{result.OverallEntropy:F4}");

            if (!string.IsNullOrEmpty(result.Reason))
                sb.Append($$",\"reason\":\"{result.Reason}\"");

            // Add chunks summary
            var highEntropyCount = result.Chunks.Count(c => c.IsHighEntropy);
            sb.Append($",\"high_entropy_chunks\":{highEntropyCount}");

            // Add peaks
            if (result.Peaks.Count > 0)
            {
                sb.Append(",\"peaks\":[");
                for (int i = 0; i < result.Peaks.Count; i++)
                {
                    var p = result.Peaks[i];
                    sb.Append('{');
                    sb.Append($"\"offset\":{p.StartOffset},");
                    sb.Append($"\"length\":{p.Length},");
                    sb.Append($"\"entropy\":{p.AverageEntropy:F2}");
                    
                    if (!string.IsNullOrEmpty(p.LikelyCause))
                        sb.Append($$",\"cause\":\"{p.LikelyCause}\"");

                    sb.Append('}');
                    if (i < result.Peaks.Count - 1) sb.Append(',');
                }
                sb.Append(']');
            }

            // Add summary stats
            var avgEntropy = result.Chunks.Count > 0 
                ? result.OverallEntropy / result.Chunks.Count 
                : 0;
            
            sb.Append($",\"avg_entropy\":{avgEntropy:F4}");

            sb.Append('}');
            return sb.ToString();
        }

        /// <summary>
        /// Outputs SARIF-compatible format for CI integration.
        /// </summary>
        public string ToSarif(AnalysisResult result)
        {
            var runs = new List<Run>();

            foreach (var chunk in result.Chunks.Where(c => c.IsHighEntropy))
            {
                var locations = new List<Location[]>();
                
                // Create location for this high-entropy region
                var locs = new[]
                {
                    new Location
                    {
                        PhysicalLocation = new PhysicalLocation
                        {
                            ArtifactLocation = new ArtifactLocation
                            {
                                Uri = $"file://{chunk.Offset:X8}"
                            }
                        },
                        Region = new Region
                        {
                            StartLine = 1,
                            EndLine = 1,
                            StartColumn = 0,
                            EndColumn = chunk.Size
                        }
                    }
                };

                var runsItem = new Run
                {
                    Name = "High Entropy Detection",
                    Tool = new Tool
                    {
                        Driver = new ToolComponent
                        {
                            Name = "EntropyAnalyzer",
                            Version = "1.0.0"
                        }
                    },
                    Results = new List<Result>
                    {
                        new Result
                        {
                            Level = chunk.Entropy > 7.5 ? "error" : "warning",
                            Message = new Message
                            {
                                Text = $"High entropy region detected: {chunk.Entropy:F2} bits/byte"
                            },
                            Locations = locs,
                            RuleId = "HIGH_ENTROPY",
                            RuleName = "Possible Packer or Encryption"
                        }
                    }
                };

                runs.Add(runsItem);
            }

            var sarif = new SarifReport
            {
                SchemaVersion = "2.1.0",
                Runs = runs,
                Version = 1
            };

            return JsonSerializer.Serialize(sarif);
        }

        /// <summary>
        /// Clears internal buffers for memory management.
        /// </summary>
        public void Clear()
        {
            _chunks.Clear();
            HighEntropyPeak = null;
        }

        /// <summary>
        /// Disposes of managed resources.
        /// </summary>
        public void Dispose()
        {
            Clear();
            GC.SuppressFinalize(this);
        }

        ~EntropyAnalyzer() => Dispose();
    }

    // ============================================================================
    // SARIF Data Models (Simplified for compatibility)
    // ============================================================================

    public class SarifReport
    {
        public string SchemaVersion { get; set; } = "2.1.0";
        public int Version { get; set; }
        public List<Run> Runs { get; set; } = new List<Run>();
    }

    public class Run
    {
        public string Name { get; set; }
        public Tool Tool { get; set; }
        public List<Result> Results { get; set; } = new List<Result>();
    }

    public class Tool
    {
        public ToolComponent Driver { get; set; }
    }

    public class ToolComponent
    {
        public string Name { get; set; }
        public string Version { get; set; }
    }

    public class Result
    {
        public string Level { get; set; } = "warning";
        public Message Message { get; set; }
        public List<Location> Locations { get; set; } = new List<Location>();
        public string RuleId { get; set; }
        public string RuleName { get; set; }
    }

    public class Message
    {
        public string Text { get; set; }
    }

    public class Location
    {
        public PhysicalLocation PhysicalLocation { get; set; }
        public Region Region { get; set; }
    }

    public class PhysicalLocation
    {
        public ArtifactLocation ArtifactLocation { get; set; }
    }

    public class ArtifactLocation
    {
        public string Uri { get; set; }
    }

    public class Region
    {
        public int StartLine { get; set; }
        public int EndLine { get; set; }
        public int StartColumn { get; set; }
        public int EndColumn { get; set; }
    }

    // ============================================================================
    // Analysis Result Container
    // ============================================================================

    public sealed class AnalysisResult
    {
        public string Name { get; set; } = "";
        public string Path { get; set; } = "";
        public long TotalSize { get; set; }
        public double OverallEntropy { get; set; }
        
        /// <summary>High entropy chunk count (entropy > 7.0)</summary>
        public int HighEntropyChunkCount => 
            Chunks.Count(c => c.Entropy > 7.0);

        /// <summary>All chunks analyzed</summary>
        public List<ChunkResult> Chunks { get; set; } = new List<ChunkResult>();

        /// <summary>Detect peak regions</summary>
        public List<PeakRegion> Peaks { get; set; } = new List<PeakRegion>();

        /// <summary>Reason for low/high entropy classification</summary>
        public string Reason { get; set; } = "";

        /// <summary>Classification: Normal, Suspicious, LikelyPacked</summary>
        public ClassificationLevel Classification => 
            Classify(OverallEntropy);

        public override string ToString()
        {
            return $"[{Name}] Size={TotalSize} bytes, " +
                   $"Entropy={OverallEntropy:F2}, " +
                   $"Classification={Classification}";
        }
    }

    ///

File: polyglot/typescript/entropy_analyzer.ts
import * as fs from 'fs';
import * as path from 'path';

// ============================================================================
// Configuration & Interfaces
// ============================================================================

export interface EntropyConfig {
  windowSize?: number;        // Bytes per window (default: 0x1000)
  stepSize?: number;          // Step between windows (default: 0x200)
  entropyThreshold?: number;  // Min entropy to flag as suspicious (default: 7.5)
  minWindows?: number;        // Minimum windows for valid report (default: 1)
  outputFormat?: 'json' | 'sarif'; // Output format
}

export interface WindowResult {
  offset: number;
  size: number;
  entropy: number;
  isSuspicious: boolean;
  byteDistribution: Map<number, number>;
}

export interface EntropyReport {
  filename: string;
  fileSize: number;
  totalWindows: number;
  avgEntropy: number;
  maxEntropy: number;
  minEntropy: number;
  suspiciousRegions: WindowResult[];
  summary: {
    highEntropyCount: number;
    highEntropyPercentage: number;
    estimatedPacked: boolean;
  };
}

export interface SarifIssue {
  id: string;
  ruleId: string;
  level: 'error' | 'warning' | 'note';
  message: string;
  locations: Array<{
    physicalLocation: {
      artifactLocation: { uri: string };
    };
  }>;
}

// ============================================================================
// Core Entropy Calculation
// ============================================================================

function calculateShannonEntropy(buffer: Buffer): number {
  const byteFreq = new Map<number, number>();
  
  // Count byte frequencies
  for (let i = 0; i < buffer.length; i++) {
    const b = buffer[i];
    byteFreq.set(b, (byteFreq.get(b) || 0) + 1);
  }

  let entropy = 0.0;
  const totalBytes = buffer.length;

  // Shannon formula: H = -Σ(p * log₂(p))
  for (const [byte, count] of byteFreq.entries()) {
    if (count > 0) {
      const p = count / totalBytes;
      entropy -= p * Math.log2(p);
    }
  }

  return entropy;
}

// ============================================================================
// Sliding Window Analysis
// ============================================================================

function analyzeWindows(
  buffer: Buffer, 
  config: EntropyConfig
): WindowResult[] {
  const windowSize = config.windowSize || 0x1000;
  const stepSize = config.stepSize || 0x200;
  
  if (buffer.length < windowSize) {
    return [];
  }

  const results: WindowResult[] = [];
  let offset = 0;

  while (offset + windowSize <= buffer.length) {
    // Extract current window
    const windowBuffer = buffer.slice(offset, offset + windowSize);
    
    // Calculate entropy for this window
    const entropy = calculateShannonEntropy(windowBuffer);
    
    // Determine if suspicious based on threshold
    const isSuspicious = entropy >= (config.entropyThreshold || 7.5);

    results.push({
      offset,
      size: windowSize,
      entropy,
      isSuspicious,
      byteDistribution: new Map([...windowBuffer.entries()]),
    });

    offset += stepSize;
  }

  return results;
}

// ============================================================================
// Report Generation
// ============================================================================

function generateReport(
  filename: string, 
  buffer: Buffer, 
  config: EntropyConfig
): EntropyReport {
  const windowSize = config.windowSize || 0x1000;
  const stepSize = config.stepSize || 0x200;
  
  if (buffer.length < windowSize) {
    return createEmptyReport(filename, buffer.length);
  }

  // Analyze all windows
  const results = analyzeWindows(buffer, config);
  
  // Calculate summary statistics
  let totalEntropy = 0.0;
  let maxEntropy = -Infinity;
  let minEntropy = Infinity;
  let highEntropyCount = 0;

  for (const result of results) {
    totalEntropy += result.entropy;
    if (result.entropy > maxEntropy) maxEntropy = result.entropy;
    if (result.entropy < minEntropy) minEntropy = result.entropy;
    
    if (result.isSuspicious) {
      highEntropyCount++;
    }
  }

  const avgEntropy = results.length > 0 ? totalEntropy / results.length : 0;
  const suspiciousRegions = results.filter(r => r.isSuspicious);
  
  // Estimate if file is packed based on overall entropy profile
  const estimatedPacked = maxEntropy >= (config.entropyThreshold || 7.5) && 
                          highEntropyCount > results.length * 0.1;

  return {
    filename,
    fileSize: buffer.length,
    totalWindows: results.length,
    avgEntropy,
    maxEntropy,
    minEntropy,
    suspiciousRegions,
    summary: {
      highEntropyCount,
      highEntropyPercentage: (highEntropyCount / results.length) * 100,
      estimatedPacked,
    },
  };
}

function createEmptyReport(filename: string, fileSize: number): EntropyReport {
  return {
    filename,
    fileSize,
    totalWindows: 0,
    avgEntropy: 0,
    maxEntropy: 0,
    minEntropy: 0,
    suspiciousRegions: [],
    summary: {
      highEntropyCount: 0,
      highEntropyPercentage: 0,
      estimatedPacked: false,
    },
  };
}

// ============================================================================
// SARIF Output Generation
// ============================================================================

function generateSarif(report: EntropyReport): string {
  const sarifVersion = '2.1.0';
  const toolName = 'EntropyAnalyzer';
  const ruleIdPrefix = 'Ent.';

  // Build issues from suspicious regions
  const issues: SarifIssue[] = report.suspiciousRegions.map((region, index) => ({
    id: `issue-${index + 1}`,
    ruleId: `${ruleIdPrefix}HIGH_ENTROPY`,
    level: 'warning',
    message: `High entropy region detected at offset ${0x${report.filename.length > 0 ? report.filename : ''}}`,
    locations: [{
      physicalLocation: {
        artifactLocation: { uri: report.filename || 'stdin' },
      },
    }],
  }));

  const sarif = {
    $schema: `http://schematics.org/sarif/${sarifVersion}`,
    version: sarifVersion,
    runs: [{
      tool: {
        name: toolName,
        version: '1.0.0',
        rules: [{
          id: `${ruleIdPrefix}HIGH_ENTROPY`,
          name: 'High Entropy Region Detected',
          description: 'Potential packed or encrypted code region',
          shortDescription: { text: 'Suspicious entropy level' },
          fullDescription: { 
            text: `Found ${report.suspiciousRegions.length} regions with entropy >= ${report.summary.estimatedPacked ? '>=' : ''}${(config.entropyThreshold || 7.5).toFixed(2)}`,
            markdown: `Found **${report.suspiciousRegions.length}** suspicious regions.\n\n| Offset | Entropy | Size |\n|--------|---------|------|\n`,
          },
        }],
      },
      results: issues,
    }],
  };

  return JSON.stringify(sarif, null, 2);
}

// ============================================================================
// File I/O Helpers
// ============================================================================

function readBufferFromFile(filePath: string): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    fs.readFile(filePath, (err, data) => {
      if (err) {
        reject(err);
        return;
      }
      resolve(data);
    });
  });
}

function readBufferFromStdin(): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];
    
    process.stdin.on('data', (chunk) => {
      chunks.push(chunk);
    });

    process.stdin.on('end', () => {
      resolve(Buffer.concat(chunks));
    });

    process.stdin.on('error', reject);
  });
}

// ============================================================================
// Main Analysis Function
// ============================================================================

export async function analyzeEntropy(
  source: string | Buffer, 
  config?: EntropyConfig
): Promise<Ent